pkg/token/daemon.go

```diff
@@ -160,6 +162,70 @@ func newDaemon(idConfig *config.IdentityConfig, tt mode) (*daemon, error) {
 	}, nil
 }
 
+// GroupDoResult contains token and its requestID after singleFlight.group.Do()
+type GroupDoResult struct {
+	requestID string
+	token     Token
+}
+
+// requestTokenToZts sends a request to ZTS server to fetch either role token or access token.
+// for mode, it only accepts mROLE_TOKEN or mACCESS_TOKEN
+// it also stores in cache for you after successful fetch.
+func (d *daemon) requestTokenToZts(k CacheKey, m mode, requestID string) (GroupDoResult, error) {
+	tokenName := "" // tokenName is used for logger (role token or access token only)
+	isRoleTokenRequested := m == mROLE_TOKEN
+	isAccessTokenRequested := m == mACCESS_TOKEN
+
+	if isRoleTokenRequested {
+		tokenName = "role token"
+	} else if isAccessTokenRequested {
+		tokenName = "access token"
+	} else {
+		return GroupDoResult{requestID: requestID, token: nil}, fmt.Errorf("Invalid mode: %d", m)
+	}
+
+	log.Debugf("Attempting to fetch %s from Athenz ZTS server: target[%s], requestID[%s]", tokenName, k.String(), requestID)
+
+	r, err, shared := d.group.Do(k.UniqueId(tokenName), func() (interface{}, error) {
+		// define variables before request to ZTS
+		var fetchedToken Token
+		var err error
+
+		if isRoleTokenRequested {
+			fetchedToken, err = fetchRoleToken(d.ztsClient, k)
+		} else { // isAccessTokenRequested
+			fetchedToken, err = fetchAccessToken(d.ztsClient, k, d.saService)
+		}
+
+		if err != nil {
+			log.Debugf("Failed to fetch %s from Athenz ZTS server: target[%s], requestID[%s]", tokenName, k.String(), requestID)
+			return GroupDoResult{requestID: requestID, token: nil}, err
+		}
+
+		if isRoleTokenRequested {
+			d.roleTokenCache.Store(k, fetchedToken)
+		} else { // isAccessTokenRequested
+			d.accessTokenCache.Store(k, fetchedToken)
+		}
+
+		log.Infof("Successfully updated %s cache: target[%s], requestID[%s]", tokenName, k.String(), requestID)
+		return GroupDoResult{requestID: requestID, token: fetchedToken}, nil
+	})
+
+	result := r.(GroupDoResult)
+	log.Debugf("requestID: [%s] handledRequestId: [%s] roleToken: [%s]", requestID, result.requestID, result.token)
```

Comment:
取得したトークンがそのままログに出てしまうので、以下のように修正した方がよいと思います
```suggestion
	log.Debugf("requestID: [%s] handledRequestId: [%s] target: [%s]", requestID, result.requestID, k.String())
```